### ISSUE TITLE
Can't start Mavlink on Iridium sat com

### ISSUE THREAD

**Issue by hugheslavigne** [PX4/Firmware#10788]
I am unable to start mavlink on the Iridium sat com device.

First, I can start the Iridium device in the QGC shell using this command:
iridiumsbd start -d /dev/ttyS2 -b 115200

Then I can send messages using the test function with s argument or at command.
However, when I try to start Mavlink with this command :
mavlink start -d /dev/iridium -b 115200 -m iridium
I get this error log:
INFO [mavlink] mode: Iridium, data rate: 5760 B/s on /dev/iridium @ 115200B
ERROR [mavlink] ERR GET CONF /dev/iridium: -1
ERROR [mavlink] could not open /dev/iridium

I can confirm that /dev/iridium exists.
I've also tried to change the sys_companion parameter, but it still doesn't work.

@acfloria have you experienced this before?

**Comment by dagar** [comment]
@hugheslavigne check the baudrate configuration of the RockBLOCK.

**Comment by hugheslavigne** [comment]
The RockBlock has a baudrate of 115200 as I am able to send message throught satellite communication using the iridiumsbd driver test function from PX4. This driver uses an harcoded value of 115200 for the baud rate.

I have no problem using the iridium driver, but I can't "attach" Mavlink to it. I am not sure to understand what should be the command to enable Mavlink communication on this device once it has been initialized.

**Comment by dagar** [comment]
Can you try this with current master? Instead of using the legacy SYS_COMPANION parameter try ISBD_CONFIG.

**Comment by hamishwillee** [comment]
@dagar You tagged this as documentation, but the [docs](https://docs.px4.io/en/advanced_features/satcom_roadblock.html#software) covers current setup via ISBD_CONFIG. Let me know if there is something specific I should be addressing.

**Comment by acfloria** [comment]
@hugheslavigne I did not experience such an issue so far. Can you maybe start up the driver in the verbose mode (iridiumsbd start -d /dev/ttyS2 -v) and provide me with a screenshot of the full startup sequence?

**Comment by hugheslavigne** [comment]
@dagar , I am working on the master version from september 21 2018. I'll try to update my firmware next week to test the new parameter (ISBD_CONFIG). I'm not sure it will change anything to the result as I already tested the same command that are in the .yaml.

@acfloria This is the complete sequence I entered in the mavlink shell :

NuttShell (NSH)
nsh> 
nsh> iridiumsbd start -d /dev/ttyS2 -v
INFO  [iridiumsbd] starting
WARN  [iridiumsbd] verbose mode ON
INFO  [iridiumsbd] opening Iridium SBD modem UART: /dev/ttyS2
INFO  [iridiumsbd] UART opened
INFO  [iridiumsbd] WRITING AT COMMAND: AT&K0
INFO  [iridiumsbd] WRITING AT COMMAND: ATE0
INFO  [iridiumsbd] read interval: 0 s
INFO  [iridiumsbd] SBD session timeout: 60 s
INFO  [iridiumsbd] SBD stack time: 0 ms
nsh> INFO  [iridiumsbd] UPDATING SIGNAL QUALITY
INFO  [iridiumsbd] WRITING AT COMMAND: AT
INFO  [iridiumsbd] WRITING AT COMMAND: AT+CSQ
INFO  [iridiumsbd] SWITCHING STATE FROM STANDBY TO SIGNAL CHECK
INFO  [iridiumsbd] SIGNAL QUALITY: 3
INFO  [iridiumsbd] SWITCHING STATE FROM SIGNAL CHECK TO STANDBY
mavlink start -d /dev/iridium -b 115200 -m iridium
INFO  [mavlink] mode: Iridium, data rate: 5760 B/s on /dev/iridium @ 115200B
ERROR [mavlink] ERR GET CONF /dev/iridium: -1

ERROR [mavlink] could not open /dev/iridium
nsh> ls /dev
/dev:
 accel0
 accel1
 baro0
 console
 gyro0
 gyro1
 icm20602_accel
 icm20602_gyro
 iridium
 led0
 mag0
 mmcsd0
 mpu9250_accel
 mpu9250_gyro
 mpu9250_mag
 mtdblock0
 mtdblock1
 null
 pipe0
 pipe1
 pwm_output0
 px4fmu
 ram0
 ttyACM0
 ttyS0
 ttyS1
 ttyS2
 ttyS3
 ttyS4
 ttyS5
 ttyS6

**Comment by hugheslavigne** [comment]
@dagar Just tested with the new master version and ISBD_CONFIG. I still have the same problem. I am able to communicate with the Iridium but I am unable to attach a mavlink process to it.

Any idea? Is this feature fully tested?

**Comment by acfloria** [comment]
@hugheslavigne 
It is tested as much as I could with the hardware available. Somehow I could not reproduce the error you described. Which hardware are you using? Did you try using another serial port? We usually use ttyS6 on the fmuv3 and never had an issue with the mavlink module there.

**Comment by jeffstreet** [comment]
@hugheslavigne: I'm seeing a similar issue. Did you find a solution?

**Comment by jeffstreet** [comment]
@acfloria: Can you please specify which hardware is working for you?

**Comment by hugheslavigne** [comment]
@jeffstreet I wasn't able to resolve this issue. As I had to submit a solution to my client before December 2018, I end up rewriting the Iridiumsbd driver so that it listens to the most important topics (position, attitude, etc.). Then I've encoded the data in csv format for the transmission. So, unfortunately, I wasn't able to make it work with QGC. But, I've created a little software to decode the received data with a server.

For your information, I used custom hardware for this project, as I was working with a platform that had to be waterproof. However, I'm pretty sure this problem wasn't related to the hardware I used.

If you ever find a solution, please let me know.

**Comment by jeffstreet** [comment]
@hugheslavigne: Bummer, but thanks for your speedy reply.

Do you think it's best to continue working on this problem by reopening this issue or posting a new one?

**Comment by jeffstreet** [comment]
@acfloria Can you please tell us the most recent firmware you've tested successfully with?

**Comment by acfloria** [comment]
I think the last time I tested it with upstream PX4 was when v1.8 was released.
What is the output you get when you start up the driver?
I am not 100% sure but you might have to switch the RX/TX ports because Rockblock has not the standard convention of naming them (https://docs.rockblock.rock7.com/docs/connectors)

**Comment by jeffstreet** [comment]
Here's the config:
PX4: 1.10.0beta (same output in 1.9.2 stable). 
ISBD_CONFIG: Telem2

After boot, here's the driver status
![Screenshot from 2019-09-25 08-34-12](https://user-images.githubusercontent.com/2422865/65602831-8b6db780-df72-11e9-82d0-e174a803b547.png)

Does this mean the driver starts up okay, but doesn't form a Mavlink connection?

Manually stop the driver, and restart in verbose mode:
![Screenshot from 2019-09-25 08-35-24](https://user-images.githubusercontent.com/2422865/65602926-b5bf7500-df72-11e9-80c0-5cf1ee2f884d.png)
![Screenshot from 2019-09-25 08-36-07](https://user-images.githubusercontent.com/2422865/65602932-b9eb9280-df72-11e9-8f56-d845b97ae5f0.png)

Does the "Signal quality: 5" at the bottom suggests the modem is communicating with the pixhawk and the modem is getting a satellite signal?

**Comment by jeffstreet** [comment]
Also tried creating an /etc/extras.txt file, per the latest [SatComInfrastructure documentation section Vehicle Setup > Software](https://github.com/acfloria/SatComInfrastructure). Output was identical to above (i.e. no Mavlink instance on /dev/ttyS2).

Tried to manually start mavlink as @hugheslavigne did, but got identical result.

Config details:
PX4: 1.10.0beta
ISBD_CONFIG: Disabled, also tried Telem2
Connected to computer via USB

**Comment by jeffstreet** [comment]
Tried 

ISBD_CONFIG: Disabled
Left /etc/extras.txt in place

MAV_1_CONFIG: Telem2
SER_TEL2_BAUD: 15200 8N1

This setup creates a mavlink instance on /dev/ttyS2 but does not send/receive any messages.
![Screenshot from 2019-09-25 09-42-46](https://user-images.githubusercontent.com/2422865/65607254-3766d100-df7a-11e9-9fe3-49c8954c56fa.png)

Maybe change MAV_1_FORWARD, _MODE, or _RATE from defaults?

Config details:
PX4: 1.10.0beta
ISBD_CONFIG: Disabled, also tried Telem2
Connected to computer via 900 MHz radio

**Comment by jeffstreet** [comment]
@acfloria @hugheslavigne 
Got it working by reverting to PX4 1.8.0 firmware (px4fmu-v3_default.px4 on Pixhawk2 hardware)

Config:

- All parameters default.
- Iridium modem connected on GPS2 [with cable shown above](https://github.com/PX4/Firmware/issues/10788#issuecomment-533283869).
- SYS_COMPANION has a value of OSD (57600, 8N1), but I didn't change that.
- /etc/extras.txt file:

```
# add a sleep here to make sure that the module is powered
usleep 200000

#driver
if iridiumsbd start -d /dev/ttyS6 -b 115200
then
	# start the iridium mavlink instance
	mavlink start -d /dev/iridium -m iridium -b 115200

	# Sleep a bit to give Mavlink time to set up
	usleep 100000
else
	echo "WARN: Iridiumsbd driver not started, reboot"
	tune_control play -m "ML<<CP4CP4CP4CP4CP4"
fi
```

**Comment by jeffstreet** [comment]
@acfloria Would you please test with PX4 firmware 1.9.2 ?

**Comment by jeffstreet** [comment]
@acfloria Do you know who's handling the iridium driver?

**Comment by acfloria** [comment]
I wrote the driver but at the moment I probably do not have the time to investigate why it is not working anymore (maybe this weekend). Since the driver was not modified I guess some changes in the mavlink module break the communication with the driver.

**Comment by jeffstreet** [comment]
Fantastic. If it would be helpful, please feel free to ask for assistance and/or testing.

**Comment by jeffstreet** [comment]
@acfloria: Would you please post here when the driver update is ready for testing?

**Comment by acfloria** [comment]
I will but it will take some time since I am now one week in a training camp. If you need it urgently you would have to have a look at it.

**Comment by matteobarbera** [comment]
Hey was there ever a follow-up on updating the driver? I've been trying to get the iridium module working, I was wondering if the mavlink and iridiumsbd module mismatch was still an issue or if I have some other problem

**Comment by jeffstreet** [comment]
Hi @acfloria,
Have you had time to touch the driver in the past few months?
Cheers,

**Comment by jeffstreet** [comment]
@matteobarbera,
I haven't seen any followup, but am trying to reach @acfloria to see how he'd like to move forward.

If you can get it to work with PX 1.8.0 ([as described above](https://github.com/PX4/Firmware/issues/10788#issuecomment-535165248)), then the driver is likely the problem. The best I could achieve intermittent operation.

**Comment by matteobarbera** [comment]
@jeffstreet I unfortunately don't have access to the RockBlock hardware anymore, as I was working on it as part of an internship I have now concluded. 

Before finishing, I was able to receive a sporadic Mavlink heartbeat on QGC from a drone running the latest stable release on a Pixhawk 4 (inspected with `iridiumsbd status` in the Mavlink Console). Data packets were being sent, as RockBlock credits were being used up. QGC did not seem to be using the data from the iridium link, however, and I was never able to make the `LinkIndicator` appear as described in the SatcomInfrastructure documentation.
